app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/RefreshController.kt

```diff
@@ -314,205 +208,143 @@ class RefreshController @Inject constructor(
 
     private suspend fun Realm.refresh(scope: CoroutineScope, folder: Folder): Set<Thread> {
 
-        val impactedThreads = mutableSetOf<Thread>()
         val previousCursor = folder.cursor
+        val impactedThreads = mutableSetOf<Thread>()
 
-        impactedThreads += if (previousCursor == null) {
-            fetchOneNewPage(scope, folder, shouldUpdateCursor = true).second
+        if (previousCursor == null) {
+            fetchOldMessagesUids(scope, folder)
         } else {
             fetchActivities(scope, folder, previousCursor)
         }
 
-        if (folder.remainingOldMessagesToFetch > 0) fetchAllOldPages(scope, folder)
+        impactedThreads += fetchAllNewPages(scope, folder.id)
+        fetchAllOldPages(scope, folder.id)
 
         return impactedThreads
     }
 
-    private suspend fun Realm.fetchOneOldPage(scope: CoroutineScope, folder: Folder) {
-        fetchOnePage(scope, folder, Direction.IN_THE_PAST, shouldUpdateCursor = false)
-    }
-
-    private suspend fun Realm.fetchOneNewPage(
-        scope: CoroutineScope,
-        folder: Folder,
-        fibonacci: Int = 1,
-        shouldUpdateCursor: Boolean = false,
-    ): Pair<Int, Set<Thread>> {
-        return fetchOnePage(scope, folder, Direction.TO_THE_FUTURE, shouldUpdateCursor, fibonacci)
-    }
+    private suspend fun Realm.fetchOnePageOfOldMessages(scope: CoroutineScope, folderId: String) {
 
-    private suspend fun Realm.fetchAllOldPages(scope: CoroutineScope, folder: Folder) {
+        var totalNewThreads = 0
+        var upToDateFolder = getUpToDateFolder(folderId)
+        var maxPagesToFetch = Utils.MAX_OLD_PAGES_TO_FETCH_TO_GET_ENOUGH_NEW_THREADS
 
-        fun remainingCount() = FolderController.getFolder(folder.id, realm = this)?.remainingOldMessagesToFetch ?: -1
+        while (
+            totalNewThreads < Utils.PAGE_SIZE / 2 &&
```

Comment:
What happens if `fetchOnePage` returns 0 ?

Reply:
Discussed in DM.